frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -255,18 +257,18 @@ public synchronized SelectionResult select(ResourceDescriptor requestResource, S
     }
 
     ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
-    if (ValueRangeUtils.getValueNumber(reusePorts) > 0) {
-      LOGGER.logInfo(
-          "select: reuse pre-selected ports: [%s]", ValueRangeUtils.toString(reusePorts));
-      optimizedRequestResource.setPortRanges(reusePorts);
-    }
     if (optimizedRequestResource.getPortNumber() > 0 && ValueRangeUtils.getValueNumber(optimizedRequestResource.getPortRanges()) <= 0) {
-      //If port is required and the portRange is not set in previous steps, allocate port ranges from all candidate nodes.
-      List<ValueRange> portRanges = selectPortsFromFilteredNodes(optimizedRequestResource);
-      LOGGER.logInfo(
-          "select: select ports from all filteredNodes  :  [%s]", ValueRangeUtils.toString(portRanges));
-      if (ValueRangeUtils.getValueNumber(portRanges) == optimizedRequestResource.getPortNumber()) {
-        optimizedRequestResource.setPortRanges(portRanges);
+      if (ValueRangeUtils.getValueNumber(reusedPorts) > 0) {
+        LOGGER.logInfo(
+            "select: reuse pre-selected ports: %s", CommonExts.toString(reusedPorts));
+        optimizedRequestResource.setPortRanges(reusedPorts);
+      } else {
+        List<ValueRange> portRanges = selectPortsFromFilteredNodes(optimizedRequestResource);
+        LOGGER.logInfo(
+            "select: select ports from all filteredNodes: %s", CommonExts.toString(portRanges));
+        if (ValueRangeUtils.getValueNumber(portRanges) >= optimizedRequestResource.getPortNumber()) {
```

Comment:
port in PortRanges must equal with getPortNubmer(), If it is >, there are some issue in selectPortsFromFilteredNodes()

Reply:
OK

---
In reply to: [189478035](https://github.com/Microsoft/pai/pull/588#discussion_r189478035) [](ancestors = 189478035)